### PR TITLE
[Merged by Bors] - feat(SetTheory/ZFC): Add `PSet.rank` and `ZFSet.rank`

### PR DESCRIPTION
**Pull request by staroperator**
Add ordinal ranks of `PSet` and `ZFSet`.

---

- [x] depends on: #16006

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    intro; exact rank_le_of_subset
+  · simp; apply rank_lt_of_mem; simp
+
+/-- For the rank of `⋃₀ x`, we only have `rank (⋃₀ x) ≤ rank x ≤ rank (⋃₀ x) + 1`.
+    This inequality is splitted into `rank_sUnion_le` and `succ_rank_sUnion_ge`. -/
+theorem rank_sUnion_le : rank (⋃₀ x : ZFSet) ≤ rank x := by
+  apply rank_le_of_forall_mem_rank_lt
+  simp; intros; trans <;> apply rank_lt_of_mem <;> assumption
+
+theorem succ_rank_sUnion_ge : rank (⋃₀ x : ZFSet) + 1 ≥ rank x := by
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:106]:
```suggestion
theorem succ_rank_sUnion_ge : rank x ≤ succ (rank (⋃₀ x : ZFSet)) := by
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal}
+    : (∀ y ∈ x, rank y < o) → rank x ≤ o :=
+  Quotient.inductionOn x fun _ h =>
+    PSet.rank_le_of_forall_mem_rank_lt fun y h' => h ⟦y⟧ h'
+
+theorem rank_le_of_subset (h : x ⊆ y) : rank x ≤ rank y :=
+  rank_le_of_forall_mem_rank_lt fun _ h₁ => rank_lt_of_mem (h h₁)
+
+theorem rank_empty : rank ∅ = 0 := by
```
[the remark this answers is not in the document]
Comment (staroperator) [reply Mathlib/SetTheory/ZFC/Rank.lean]: It seems that `PSet` lacks some basic lemmas, e.g. `mem_insert_iff`. I can add them into SetTheory/ZFC/Basic.lean, but I'm not sure whether those changes should be put in this PR or a new PR.

[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Rank.lean]: That should go in a new PR.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal} :
+    (∀ y ∈ x, rank y < o) → rank x ≤ o :=
+  Quotient.inductionOn x fun _ h =>
+    PSet.rank_le_of_forall_mem_rank_lt fun y h' => h ⟦y⟧ h'
+
+theorem rank_le_of_subset (h : x ⊆ y) : rank x ≤ rank y :=
+  rank_le_of_forall_mem_rank_lt fun _ h₁ => rank_lt_of_mem (h h₁)
+
+@[simp] theorem rank_empty : rank ∅ = 0 := by
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:71]:
```suggestion
@[simp]
theorem rank_empty : rank ∅ = 0 := by
```
and throughout.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  rw [← Ordinal.le_zero]; apply rank_le_of_forall_mem_rank_lt; simp
+
+@[simp] theorem rank_insert : rank (insert x y) = max (succ (rank x)) (rank y) := by
+  apply le_antisymm
+  · apply rank_le_of_forall_mem_rank_lt; simp
+    intro _ h; exact Or.inr (rank_lt_of_mem h)
+  · simp; constructor
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:78]: Using `simp` like this when it's not at the end of a proof is generally discouraged. If the set of applicable simp lemmas changes, the proof might suddenly and unexpectedly break.

You can use `simp?` to figure out what theorems are being called, and then use `simp only` with the theorems that are required. I think you'll find that in most cases, your `simp` can be replaced with a simple `rw` sequence.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+end PSet
+
+namespace ZFSet
+
+variable {x y : ZFSet.{u}}
+
+/-- The ordinal rank of a ZFC set -/
+noncomputable def rank : ZFSet.{u} → Ordinal.{u} :=
+  Quotient.lift PSet.rank PSet.rank_eq_of_equiv
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:58]:
```suggestion
  Quotient.lift _ PSet.rank_eq_of_equiv
```
This is a somewhat cheeky golf, so feel free to ignore.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+theorem rank_lt_of_mem : {x y : PSet} → y ∈ x → rank y < rank x
+  | ⟨_, _⟩, _, ⟨_, h⟩ => by
+    rw [rank_eq_of_equiv _ _ h]
+    apply lt_lsub
+
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal} :
+    {x : PSet} → (∀ y ∈ x, rank y < o) → rank x ≤ o
+  | ⟨_, A⟩, h => lsub_le fun a => h (A a) (Mem.mk A a)
+
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:49]:
```lean 4
theorem rank_le_iff {o : Ordinal} {x : PSet} : rank x ≤ o ↔ ∀ y ∈ x, rank y < o :=
  ⟨fun hx _ hy => (rank_lt_of_mem hy).trans_le hx, rank_le_of_forall_mem_rank_lt⟩
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal} :
+    (∀ y ∈ x, rank y < o) → rank x ≤ o :=
+  Quotient.inductionOn x fun _ h =>
+    PSet.rank_le_of_forall_mem_rank_lt fun y h' => h ⟦y⟧ h'
+
+theorem rank_le_of_subset (h : x ⊆ y) : rank x ≤ rank y :=
+  rank_le_of_forall_mem_rank_lt fun _ h₁ => rank_lt_of_mem (h h₁)
+
+@[simp] theorem rank_empty : rank ∅ = 0 := by
+  rw [← Ordinal.le_zero]; apply rank_le_of_forall_mem_rank_lt; simp
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:72]:
```suggestion
  rw [← Ordinal.le_zero, rank_le_iff, forall_mem_empty]
```
The last two lemmas are missing, but would be nice to have.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  · apply rank_le_of_forall_mem_rank_lt; simp
+    intro _ h; exact Or.inr (rank_lt_of_mem h)
+  · simp; constructor
+    · apply rank_lt_of_mem; simp
+    · apply rank_le_of_subset; intro; apply mem_insert_of_mem
+
+@[simp] theorem rank_singleton : rank {x} = succ (rank x) :=
+  rank_insert.trans (by simp)
+
+theorem rank_pair : rank {x, y} = succ (max (rank x) (rank y)) := by
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:85]:
```suggestion
theorem rank_pair : rank {x, y} = max (succ (rank x)) (succ (rank y)) := by
  simp
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+theorem succ_rank_sUnion_ge : rank x ≤ succ (rank (⋃₀ x : ZFSet)) := by
+  rw [← rank_powerset]
+  apply rank_le_of_subset
+  intro z _; simp; intro _ _; simp; exists z
+
+@[simp] theorem rank_range {α : Type u} {f : α → ZFSet.{max u v}} :
+    rank (range f) = lsub fun i => rank (f i) := by
+  apply le_antisymm
+  · apply rank_le_of_forall_mem_rank_lt; simp; apply lt_lsub
+  · apply lsub_le; intro; apply rank_lt_of_mem; simp
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:115]:
```suggestion
```suggestion
@[simp]
theorem rank_range {α : Type u} {f : α → ZFSet.{max u v}} :
    rank (range f) = lsub fun i => rank (f i) := by
  apply (lsub_le _).antisymm'
  · apply rank_le_of_forall_mem_rank_lt
    simpa using lt_lsub _
  · intro
    apply rank_lt_of_mem
    simp
```
We seem to be missing `mem_range_self`, that would take care of the bottom `simp`.

```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+/-- `ZFSet.rank` is equal to the `WellFounded.rank` over `∈`. -/
+theorem rank_eq_wf_rank : lift.{u + 1, u} (rank x) = mem_wf.rank x := by
```
Comment (vihdzp) [inline Mathlib/SetTheory/ZFC/Rank.lean:118]:
```suggestion
theorem rank_eq_wf_rank : lift.{u + 1, u} (rank x) = mem_wf.rank x := by
  induction' x using inductionOn with x ih
  rw [mem_wf.rank_eq]
  simp_rw [← fun y : { y // y ∈ x } => ih y y.2]
  apply (le_of_forall_lt _).antisymm (Ordinal.sup_le _) <;> intro h
  · rw [lt_lift_iff]
    rintro ⟨o, rfl, h⟩
    rw [Ordinal.lt_sup]
    simpa using rank_lt_iff.2 h
  · simpa using rank_lt_of_mem h.2
```
`rank_lt_iff` would be the converse of my proposed `rank_le_iff` earlier.

File: Mathlib/SetTheory/ZFC/Rank.lean (hunk 7 above)
Comment (staroperator) [reply Mathlib/SetTheory/ZFC/Rank.lean:73]: `simp only [not_mem_empty, false_implies, implies_true]` (or simply `simp`) can solve for `forall_mem_empty`. I think it's fine to leave a `simp` at the end of proof here.

File: Mathlib/SetTheory/ZFC/Rank.lean (hunk 9 above)
Comment (staroperator) [reply Mathlib/SetTheory/ZFC/Rank.lean]: `simp only [mem_range, Set.mem_range, exists_apply_eq_apply]` already works 😄

**Review by vihdzp — APPROVED** [review]
Save for the few semicolons here and there, LGTM.

File: Mathlib/SetTheory/ZFC/Rank.lean (hunk 7 above)
Comment (vihdzp) [reply Mathlib/SetTheory/ZFC/Rank.lean]: Sure, but this is shorter and still avoids the `simp`.

File: Mathlib/SetTheory/ZFC/Basic.lean [hunk 11]
```diff
@@ -232,6 +232,12 @@ theorem equiv_iff_mem {x y : PSet.{u}} : Equiv x y ↔ ∀ {w : PSet.{u}}, w ∈
 theorem Mem.congr_left : ∀ {x y : PSet.{u}}, Equiv x y → ∀ {w : PSet.{u}}, x ∈ w ↔ y ∈ w
   | _, _, h, ⟨_, _⟩ => ⟨fun ⟨a, ha⟩ => ⟨a, h.symm.trans ha⟩, fun ⟨a, ha⟩ => ⟨a, h.trans ha⟩⟩
 
+theorem mem_of_subset {x y z : PSet} : x ⊆ y → z ∈ x → z ∈ y
+  | h₁, ⟨a, h₂⟩ => (h₁ a).elim fun b h₃ => ⟨b, h₂.trans h₃⟩
+
+theorem subset_iff {x y : PSet} : x ⊆ y ↔ ∀ ⦃z⦄, z ∈ x → z ∈ y :=
+  ⟨fun h _ => mem_of_subset h, fun h a => h (Mem.mk _ a)⟩
+
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Basic.lean:240]: These lemmas already exist on master. Can you merge master to make them disappear from the diff?

Comment (staroperator) [reply Mathlib/SetTheory/ZFC/Basic.lean]: Thanks! I have merged.

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+open Ordinal Order
+
+namespace PSet
+
+/-- The ordinal rank of a pre-set -/
+noncomputable def rank : PSet.{u} → Ordinal.{u}
+  | ⟨_, A⟩ => lsub fun a => rank (A a)
+
+theorem rank_eq_of_equiv : {x y : PSet} → Equiv x y → rank x = rank y
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:32]:
```suggestion
theorem rank_congr : ∀ {x y : PSet}, Equiv x y → rank x = rank y
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+      · obtain ⟨b, h'⟩ := βα a
+        exists b
+        rw [← h, rank_eq_of_equiv h'])
+
+theorem rank_lt_of_mem : {x y : PSet} → y ∈ x → rank y < rank x
+  | ⟨_, _⟩, _, ⟨_, h⟩ => by
+    rw [rank_eq_of_equiv h]
+    apply lt_lsub
+
+theorem rank_le_iff {o : Ordinal} : {x : PSet} → rank x ≤ o ↔ ∀ y ∈ x, rank y < o
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:49]:
```suggestion
theorem rank_le_iff {o : Ordinal} : {x : PSet} → rank x ≤ o ↔ ∀ ⦃y⦄, y ∈ x → rank y < o
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    ⟨fun h _ h' => (rank_lt_of_mem h').trans_le h, fun h =>
+      lsub_le fun a => h (A a) (Mem.mk A a)⟩
+
+theorem lt_rank_iff {o : Ordinal} {x : PSet} : o < rank x ↔ ∃ y ∈ x, o ≤ rank y := by
+  rw [← not_iff_not, not_lt, rank_le_iff]
+  simp
+
+variable {x y : PSet.{u}}
+
+theorem rank_mono (h : x ⊆ y) : rank x ≤ rank y :=
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:60]:
```suggestion
@[gcongr] theorem rank_mono (h : x ⊆ y) : rank x ≤ rank y :=
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+  apply le_antisymm
+  · simp_rw [rank_le_iff, mem_powerset, lt_succ_iff]
+    intro
+    exact rank_mono
+  · rw [succ_le_iff]
+    apply rank_lt_of_mem
+    simp
+
+/-- `PSet.rank` is equal to the `WellFounded.rank` over `∈`. -/
+theorem rank_eq_wf_rank : lift.{u + 1, u} (rank x) = mem_wf.rank x := by
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:95]:
```suggestion
theorem rank_eq_wfRank : lift.{u + 1, u} (rank x) = mem_wf.rank x := by
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+variable {x y : ZFSet.{u}}
+
+/-- The ordinal rank of a ZFC set -/
+noncomputable def rank : ZFSet.{u} → Ordinal.{u} :=
+  Quotient.lift _ fun _ _ => PSet.rank_eq_of_equiv
+
+theorem rank_lt_of_mem : y ∈ x → rank y < rank x :=
+  Quotient.inductionOn₂ x y fun _ _ => PSet.rank_lt_of_mem
+
+theorem rank_le_iff {o : Ordinal} : rank x ≤ o ↔ (∀ y ∈ x, rank y < o) :=
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:118]:
```suggestion
theorem rank_le_iff {o : Ordinal} : rank x ≤ o ↔ ∀ ⦃y⦄, y ∈ x → rank y < o :=
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+  ⟨fun h _ h' => (rank_lt_of_mem h').trans_le h,
+    Quotient.inductionOn x fun _ h =>
+      PSet.rank_le_iff.2 fun y h' => h ⟦y⟧ h'⟩
+
+theorem lt_rank_iff {o : Ordinal} : o < rank x ↔ ∃ y ∈ x, o ≤ rank y := by
+  rw [← not_iff_not, not_lt, rank_le_iff]
+  simp
+
+theorem rank_mono (h : x ⊆ y) : rank x ≤ rank y :=
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:127]:
```suggestion
@[gcongr] theorem rank_mono (h : x ⊆ y) : rank x ≤ rank y :=
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+  Quotient.inductionOn x fun _ => PSet.rank_powerset
+
+/-- For the rank of `⋃₀ x`, we only have `rank (⋃₀ x) ≤ rank x ≤ rank (⋃₀ x) + 1`.
+    This inequality is splitted into `rank_sUnion_le` and `succ_rank_sUnion_ge`. -/
+theorem rank_sUnion_le : rank (⋃₀ x) ≤ rank x := by
+  simp_rw [rank_le_iff, mem_sUnion]
+  intro _ ⟨_, _, _⟩
+  trans <;> apply rank_lt_of_mem <;> assumption
+
+theorem succ_rank_sUnion_ge : rank x ≤ succ (rank (⋃₀ x)) := by
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:163]: Don't use `ge` either in lemma statements or lemma names as people will be looking for `le` instead
```suggestion
theorem le_succ_rank_sUnion : rank x ≤ succ (rank (⋃₀ x)) := by
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    intro
+    apply Or.imp <;> apply rank_lt_of_mem
+  · apply max_le <;> apply rank_mono <;> intro _ h <;> simp [h]
+
+@[simp]
+theorem rank_powerset : rank (powerset x) = succ (rank x) :=
+  Quotient.inductionOn x fun _ => PSet.rank_powerset
+
+/-- For the rank of `⋃₀ x`, we only have `rank (⋃₀ x) ≤ rank x ≤ rank (⋃₀ x) + 1`.
+    This inequality is splitted into `rank_sUnion_le` and `succ_rank_sUnion_ge`. -/
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:157]:
```suggestion

This inequality is split into `rank_sUnion_le` and `succ_rank_sUnion_ge`. -/
```

File: Mathlib/SetTheory/ZFC/Rank.lean [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+  intro _ _
+  rw [mem_sUnion]
+  exists z
+
+@[simp]
+theorem rank_range {α : Type u} {f : α → ZFSet.{max u v}} :
+    rank (range f) = lsub fun i => rank (f i) := by
+  apply (lsub_le _).antisymm'
+  · simpa [rank_le_iff] using lt_lsub _
+  · simp [rank_lt_of_mem]
```
Comment (YaelDillies) [inline Mathlib/SetTheory/ZFC/Rank.lean:177]: These lemmas are also statable and true for `PSet`, right? Can you do that?

**Review by YaelDillies — COMMENTED** [review]
Thanks! This is good quality code. Only minor comments below.

File: Mathlib/SetTheory/ZFC/Rank.lean (hunk 20 above)
Comment (staroperator) [reply Mathlib/SetTheory/ZFC/Rank.lean]: There is no `range` for `PSet`. For `sUnion`, their theorems have been added now.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**